### PR TITLE
feat: dependant toggles

### PR DESCRIPTION
**Pull request by gardleopard**
This pr implements dependant feature toggles. The tests are specified in the client specifications https://github.com/Unleash/client-specification/pull/63

### CODE REVIEW

File: lib/unleash/feature_toggle.rb [hunk 1]
```diff
@@ -75,9 +76,47 @@ def am_enabled?(context)
...
+      end
+
+      unless parent["variants"].nil? || parent["variants"].empty?
+        return parent["variants"].include?(parent_toggle.get_variant(context).name)
+      end
+      return evaluation_result
+    end
+
+    def get_parent(feature)
+      toggle_as_hash = Unleash&.toggles&.select{ |toggle| toggle['name'] == feature }&.first
```
Comment (kwasniew) [inline lib/unleash/feature_toggle.rb:106]: is there Array.find method that performs select and first?

File: lib/unleash/feature_toggle.rb [hunk 2]
```diff
@@ -75,9 +76,47 @@ def am_enabled?(context)
       evaluate(context).enabled?
     end
 
+    def parent_dependencies_satisfied?(context)
```
Comment (kwasniew) [inline lib/unleash/feature_toggle.rb:79]:
```ruby
def parent_dependencies_satisfied?(context)
  dependencies.empty? || dependencies.all? { |parent| evaluate_parent(parent, context) }
end
```

File: lib/unleash/feature_toggle.rb [hunk 3]
```diff
@@ -75,9 +76,47 @@ def am_enabled?(context)
...
+      return true if dependencies.empty?
+      return dependencies.all? do |parent|
+        evaluate_parenthood(parent, context)
+      end
+    end
+
+    def evaluate_parenthood(parent, context)
```
Comment (kwasniew) [inline lib/unleash/feature_toggle.rb:86]:
```ruby
def evaluate_parent(parent, context)
  parent_toggle = get_parent(parent["feature"])
  return false if parent_toggle.nil? || !parent_toggle.dependencies.empty?

  evaluation_result = parent_toggle.is_enabled?(context)
  if parent["enabled"] == false
    return !evaluation_result
  elsif !evaluation_result
    return false
  end

  return evaluation_result if parent["variants"].nil? || parent["variants"].empty?

  parent["variants"].include?(parent_toggle.get_variant(context).name)
end
```

**Review by kwasniew — COMMENTED** [review]
Added some minor comments

File: lib/unleash/feature_toggle.rb (hunk 1 above)
Comment (gardleopard) [reply lib/unleash/feature_toggle.rb:107]: yes

Comment (gardleopard) [reply lib/unleash/feature_toggle.rb]: fixed

File: lib/unleash/feature_toggle.rb [hunk 4]
```diff
@@ -75,9 +76,48 @@ def am_enabled?(context)
...
+      parent_toggle = get_parent(parent["feature"])
+
+      return false if parent_toggle.nil?
+
+      return false unless parent_toggle.dependencies.empty?
+
+      evaluation_result = parent_toggle.is_enabled?(context)
+      return !evaluation_result if parent["enabled"] == false
```
Comment (rarruda) [inline lib/unleash/feature_toggle.rb:95]: nit:
```suggestion
      return !evaluation_result unless parent["enabled"]
```

File: lib/unleash/feature_toggle.rb (hunk 2 above)
Comment (gardleopard) [reply lib/unleash/feature_toggle.rb:79]: Thanks, fixed

File: lib/unleash/feature_toggle.rb (hunk 4 above)
Comment (rarruda) [inline lib/unleash/feature_toggle.rb:91]:
```suggestion
      return !parent_toggle.is_enabled?(context) unless parent["enabled"]
```

Comment (gardleopard) [reply lib/unleash/feature_toggle.rb:88]: evaluation_result is used throughout the function several times and parent.enabled cant be used like this in the code.

**Review by kwasniew — APPROVED** [review]

**Review by sighphyre — CHANGES_REQUESTED** [review]
What happens to metrics? Parent features shouldn't get metrics incremented. Can we add a test for this?

**Review by sighphyre — APPROVED** [review]
G. I realise we want tests for metrics, this is quite tough to do with current structure. Yggdrasil already has those covered. If we want to move to that faster in lieu of tests here that's.... probably okay

**Comment by gardleopard** [comment]
> What happens to metrics? Parent features shouldn't get metrics incremented. Can we add a test for this?

We are not touching the path of metrics here. Metrics behaviour is not tested in the sdk at all. If we want to have test for metrics behaviour I vote for having that as a separate PR.

**Comment by kwasniew** [comment]
@gardleopard the important part is that introducing dependent features shouldn't trigger metrics reported on the parent feature. If that's the case then we're fine. Tests can be added separately.

**Pull request merged**